src/fsharp/FSharp.Core/seq.fs

```diff
@@ -842,6 +797,486 @@ namespace Microsoft.FSharp.Collections
+
+            and SkipThenTake<'T> (startIdx:int, endIdx:int) =
+                inherit SeqComponent<'T,'T>()
+
+                let mutable count = 0
+
+                override __.ProcessNext (input:'T, halted:byref<bool>, output:byref<'T>) : bool = 
+                    if count < startIdx then
+                        count <- count + 1
+                        false
+                    elif count < endIdx then
+                        count <- count + 1
+                        output <- input
+                        halted <- count = endIdx
+                        true
+                    else
+                        halted <- true
+                        false
+
+                override __.OnComplete () =
+                    if count < startIdx then
+                        let x = startIdx - count
+                        invalidOpFmt "tried to skip {0} {1} past the end of the seq"
+                          [|SR.GetString SR.notEnoughElements; x; (if x=1 then "element" else "elements")|]
+                    elif count < endIdx then
+                        let x = endIdx - count
+                        invalidOpFmt "tried to take {0} {1} past the end of the seq"
+                            [|SR.GetString SR.notEnoughElements; x; (if x=1 then "element" else "elements")|]
+            
+            and SkipWhile<'T> (predicate: 'T -> bool) =
+                inherit SeqComponent<'T,'T>()
+
+                let mutable skip = true
+
+                override __.ProcessNext (input:'T, halted:byref<bool>, output:byref<'T>) : bool = 
+                    if skip then
+                        skip <- predicate input
+                        if skip then
+                            false
+                        else
+                            output <- input
+                            true
+                    else
+                        output <- input
+                        true
+
+            and TakeWhile<'T> (predicate: 'T -> bool) =
+                inherit SeqComponent<'T,'T>()
+
+                override __.ProcessNext (input:'T, halted:byref<bool>, output:byref<'T>) : bool = 
+                    if predicate input then
+                        output <- input
+                        true
+                    else
+                        halted <- true
+                        false
+
+            and Choose<'T, 'U> (choose:'T->'U option) =
+                inherit SeqComponent<'T,'U>()
+
+                override __.ProcessNext (input:'T, halted:byref<bool>, output:byref<'U>) : bool =
+                    match choose input with
+                    | Some value -> output <- value
+                                    true
+                    | None -> false
+
+            type SeqProcessNextStates =
+            | NotStarted = 1
+            | Finished   = 2
+            | InProcess  = 3
+
+            type SeqComposedEnumerator<'T,'U>(enumerator:IEnumerator<'T>, t2u:SeqComponent<'T,'U>) =
```

Comment:
So .... this is a reference type ... have you considered a struct enumerator, it might reduce the gc somewhat?